### PR TITLE
Gap cam

### PR DESCRIPTION
**Pull request by kretes**
This PR introduces class activation mappings from global average pooling. Whole approach is described in http://cnnlocalization.csail.mit.edu/ http://arxiv.org/pdf/1512.04150.pdf

Changes consist mostly of code that will be activated only if the selected layer is Global Average Pooling (previously quiver was throwing an exception there).

There is slightly different logic as to how to generate the 'deprocess' the image in this layer, because it is needed to compare images for various classes, and so the normalization is done only once and for whole set of visualizations, not each image on its own.

### CODE REVIEW

**Comment by kretes** [comment]
screenshot from 2017-07-08 23-32-38
This is an example of how it look like.

For GAP layer it shows that many images as there are classes to predict after it. 
UI would benefit from indicating which image is a representation of each class - it is done in the same order as the classes exist in the model, but it is not expressed anywhere in the GUI. Furhtermore - the predictions are sorted from the highest probability, which may confuse even more.

Actually for a GAP network quiver should show predictions and Class activation mappings together , so that it is clear which is which.

**Comment by jakebian** [comment]
Wow, great work. Will review tomorrow.

File: quiver_engine/vis_utils.py [hunk 1]
```diff
@@ -7,11 +7,48 @@
 def save_layer_outputs(input_img, model, layer_name, temp_folder, input_path):
 
     with get_evaluation_context():
-        layer_outputs = get_outputs_generator(model, layer_name)(input_img)[0]
 
-        if K.backend() == 'theano':
```
Comment (jakebian) [inline quiver_engine/vis_utils.py]: Why is this removed?

File: quiver_engine/vis_utils.py [hunk 2]
```diff
@@ -7,11 +7,48 @@
...
+                next_layer_weights = next_layer.get_weights()[0]
+                prev_layer_channels = next_layer_weights.shape[0]
+                num_classes = next_layer_weights.shape[1]
+                classes_outputs = np.zeros((prev_layer_outputs.shape[0],prev_layer_outputs.shape[1],len(classes)))
+                for cl in range(0,num_classes):
+                    for channel in range(0, prev_layer_channels):
+                        classes_outputs[:, :, cl] = classes_outputs[:, :, cl] + prev_layer_outputs[:,:,channel] * next_layer_weights[channel,cl]
+
+                classes_outputs = deprocess_image(classes_outputs)
+                return [save_layer_img(
```
Comment (jakebian) [inline quiver_engine/vis_utils.py]: These images for this new visualization should either have a new naming pattern or be in a subfolder

**Review by jakebian — COMMENTED** [review]
Had a question & a comment.

This is great but I think we need a way in the UI to toggle between this visualization (when available for a layer) and the naive one that's already there. I don't want to make you dig into the UI code (unless you want to), but I can merge this onto a branch and add that control before merging to master.

**Comment by kretes** [comment]
Regarding the 'toggling' - We can do that, sure. However before that change quiver didn't show anything for a 'globalaveragepooling' layer - in the logs there was an error. Since output of GAP layer is a 1-dim vector, id didn't have a shape[2] which is expected in save_layer_outputs.

File: quiver_engine/vis_utils.py (hunk 1 above)
Comment (kretes) [inline quiver_engine/vis_utils.py]: I moved and encapsulated that inside 'get_outputs_generator'

File: quiver_engine/vis_utils.py (hunk 2 above)
Comment (kretes) [inline quiver_engine/vis_utils.py]: If we want them not to be in conflict with others for the same layer - yes. I can do that in a spare moment.

**Comment by jakebian** [comment]
Ah I see, makes sense. Still the vision has always been to support many different visualizations per layer, so this is a good opportunity to start implementing that in a somewhat generic way, with the subfolder/naming, and the client side changes I'm going to make.
